test_quality_of_service/test/qos_utilities.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <tuple>
+
+#include "test_quality_of_service/qos_utilities.hpp"
+
+
+std::tuple<size_t, size_t> chrono_milliseconds_to_size_t(
+  const std::chrono::milliseconds & milliseconds)
+{
+  size_t seconds = milliseconds.count() / 1000;
+  size_t nanoseconds = (milliseconds.count() % 1000) * 1000000;
+  return std::make_tuple(seconds, nanoseconds);
+}
+
+Publisher::Publisher(
+  const std::string name,
+  const std::string topic,
+  rclcpp::PublisherOptions<> pub_options,
+  std::chrono::milliseconds publish_period)
+: Node(name),
+  name_(name),
+  topic_(topic),
+  pub_options_(pub_options),
+  publish_period_(publish_period)
+{
+  RCLCPP_INFO(this->get_logger(), "created publisher %s %s\n", name.c_str(), topic.c_str());
+  publisher_ = this->create_publisher<std_msgs::msg::String>(topic_, pub_options_);
+}
+
+void Publisher::start_publishing()
+{
+  toggle_publish_mutex.lock();
+
+  if (is_publishing) {
+    return;
+  }
+
+  is_publishing = true;
+
+  auto timer_callback =
+    [this]() -> void {
+      auto message = std_msgs::msg::String();
+
+      message.data = this->name_ + ": testing " + std::to_string(this->sent_message_count_++);
+
+      RCLCPP_INFO(this->get_logger(), this->name_ + ": publishing '%s'", message.data.c_str());
+
+      this->publisher_->publish(message);
+    };
+  timer_ = this->create_wall_timer(publish_period_, timer_callback);
+
+  toggle_publish_mutex.unlock();
+}
+
+void Publisher::stop_publishing()
+{
+  toggle_publish_mutex.lock();
+  is_publishing = false;
+
+  if (timer_) {
+    timer_->cancel();
+  }
+  toggle_publish_mutex.unlock();
+}
+
+void Publisher::toggle_publishing()
+{
+  toggle_publish_mutex.lock();
+  if (is_publishing) {
+    stop_publishing();
+  } else {
+    start_publishing();
+  }
+  toggle_publish_mutex.unlock();
+}
+
+Publisher::~Publisher()
```

Comment:
Destructor should *never* contain code that can fail.

Is it possible for cancel() to fail? If yes, it should not be in the dtor.

I don't know the code enough but it seems to be a code smell you're using a code with a separate "cleanup" function (`timer_->cancel()`) but this difference is lost at this level of abstraction. Is timer cancelling on dtor? If not, do we know why?

Reply:
> Destructor should _never_ contain code that can fail.
> 
> Is it possible for cancel() to fail? If yes, it should not be in the dtor.
> 

Why not simply catch specific exceptions in the destructor when attempting to cleanup>? At that point, if anything actually failed there hasn't been checking to account for such and destruction will continue. Doing so when freeing resources at least attempts to correct a user's mistake.

```
Publisher::~Publisher()
  try {
    // cleanup after ourselves just in case someone didn't do it properly
    if (timer_) {
        timer_->cancel();
    }
  } catch (const runtime_error& error) {
    // could try something, but seems unlikely at this point
  }
```

> I don't know the code enough but it seems to be a code smell you're using a code with a separate "cleanup" function (`timer_->cancel()`) but this difference is lost at this level of abstraction. Is timer cancelling on dtor? If not, do we know why?

I'm not following. There is no separate cleanup, as it was all done in the destructor. The timer should have been cancelled earlier by the user or other method, but this at least makes a best effort to do so if left running. Make sense?